File: python/day7.py
from util import load


def parse_lines(it, cur_dir):
    while True:
        line = next(it, None)
        if line:
            token = line.split(' ')
            if '$' == token[0]:
                if 'cd' == token[1]:
                    if '..' == token[2]:
                        break
                    else:
                        parse_lines(it, cur_dir[token[2]][1])
                elif 'ls' == token[1]:
                    pass  # ignore
            elif 'dir' == token[0]:
                cur_dir[token[1]] = ['D', {}]
            else:
                cur_dir[token[1]] = ['F', int(token[0])]
        else:
            break


def parse_input(lines):
    fs = {'/': ['D', {}]}  # initialize with root
    it = iter(lines)
    parse_lines(it, fs)
    return fs


def calculate_size(fs, limit):
    total_sz = 0
    this_sz = 0
    for f in fs.values():
        if f[0] == 'F':
            this_sz += f[1]
        else:  # it's a 'D'
            total, sub = calculate_size(f[1], limit)
            f.append(sub)
            total_sz += total + (sub if sub <= limit else 0)
            this_sz += sub
    return total_sz, this_sz


def find_to_delete(fs, at_least, current_sz):
    for f in fs.values():
        if f[0] == 'D':
            sz = f[2]
            if at_least <= sz < current_sz:
                current_sz = sz
            current_sz = find_to_delete(f[1], at_least, current_sz)
    return current_sz


def part1(fs):
    sz_limit = 100000
    return calculate_size(fs, sz_limit)[0]


def part2(fs):
    disk_sz = 70000000
    free_sz_required = 30000000
    used_sz = fs['/'][2]
    assert used_sz == 43441553
    min_delete_sz = free_sz_required - (disk_sz - used_sz)
    return find_to_delete(fs, min_delete_sz, disk_sz)


def main():
    # lines = load('day7-test.txt')
    # sz = part1(parse_input(lines))
    # assert sz == 95437

    lines = load('day7.txt')
    fs = parse_input(lines)
    sz = part1(fs)
    print(f'Total size of -100K directories: {sz}')
    assert sz == 1844187

    sz = part2(fs)
    print(f'Freeing up a directory of size: {sz}')
    assert sz == 4978279


if __name__ == "__main__":
    main()

File: rust/src/day7.rs
use std::{cell::Cell, collections::HashMap};

use crate::util::load;

#[derive(Debug)]
enum File {
    Plain(usize),
    Folder(HashMap<String, File>, Cell<usize>),
}

fn add_to_folder(folder: &mut File, name: String, entry: File) {
    match folder {
        File::Plain(_) => panic!("cannot add to plain file"),
        File::Folder(ref mut contents, _) => drop(contents.insert(name, entry)),
    }
}

fn parse_lines(it: &mut impl Iterator<Item = String>, folder: &mut File) {
    loop {
        match &it.next() {
            None => return,
            Some(line) => {
                let token: Vec<&str> = line.split(' ').collect();
                match token[0] {
                    "$" => match token[1] {
                        "cd" => match token[2] {
                            ".." => return,
                            name => {
                                if let File::Folder(ref mut map, _) = folder {
                                    if let Some(f) = map.get_mut(name) {
                                        parse_lines(it, f);
                                    }
                                }
                            }
                        },
                        "ls" => (), // nop
                        _ => panic!("unexpected cmd"),
                    },
                    "dir" => add_to_folder(
                        folder,
                        token[1].to_owned(),
                        File::Folder(HashMap::new(), Cell::new(0)),
                    ),
                    sz => add_to_folder(
                        folder,
                        token[1].to_owned(),
                        File::Plain(sz.parse::<usize>().unwrap()),
                    ),
                }
            }
        }
    }
}

fn parse_input() -> File {
    let lines: Vec<String> = load("data/day7.txt");
    let mut fs = File::Folder(HashMap::new(), Cell::new(0));
    parse_lines(&mut lines.into_iter(), &mut fs);
    fs
}

fn calculate_size(fs: &File, limit: usize) -> (usize, usize) {
    let mut total_sz = 0;
    let mut this_sz = 0;
    if let File::Folder(contents, cell) = fs {
        for f in contents.values() {
            match f {
                File::Plain(sz) => this_sz += sz,
                File::Folder(_, _) => {
                    let (total, sub) = calculate_size(f, limit);
                    total_sz += total + (if sub <= limit { sub } else { 0 });
                    this_sz += sub;
                }
            }
        }
        cell.set(this_sz);
    }
    (total_sz, this_sz)
}

fn find_to_delete(
    contents: &HashMap<String, File>,
    at_least: usize,
    mut current_sz: usize,
) -> usize {
    for f in contents.values() {
        if let File::Folder(sub_contents, cell) = f {
            let sz = cell.get();
            if at_least <= sz && sz < current_sz {
                current_sz = sz;
            }
            current_sz = find_to_delete(sub_contents, at_least, current_sz);
        }
    }
    current_sz
}

pub fn part1() -> usize {
    let sz_limit = 100000;
    let fs = parse_input();
    calculate_size(&fs, sz_limit).0
}

pub fn part2() -> usize {
    let sz_limit = 100000;
    let fs = parse_input();
    calculate_size(&fs, sz_limit);
    let disk_sz = 70000000;
    let free_sz_required = 30000000;
    match fs {
        File::Plain(_) => panic!("root folder should not be plain file"),
        File::Folder(contents, sz) => {
            let disk_used_sz = sz.get();
            let min_delete_sz = free_sz_required - (disk_sz - disk_used_sz);
            find_to_delete(&contents, min_delete_sz, disk_sz)
        }
    }
}

mod tests {
    #[test]
    fn test_part1() {
        let sz = super::part1();
        println!("Total size of -100K directories: {}", sz);
        assert_eq!(sz, 1844187);
    }

    #[test]
    fn test_part2() {
        let sz = super::part2();
        println!("Freeing up a directory of size: {}", sz);
        assert_eq!(sz, 4978279);
    }
}
